src/ShellyRGBW2/shelly_init.cpp

```diff
@@ -41,35 +44,96 @@ void CreatePeripherals(std::vector<std::unique_ptr<Input>> *inputs,
+      (struct mgos_config_lb *) mgos_sys_config_get_lb1(),
+      (struct mgos_config_lb *) mgos_sys_config_get_lb2(),
+      (struct mgos_config_lb *) mgos_sys_config_get_lb3(),
+      (struct mgos_config_lb *) mgos_sys_config_get_lb4(),
+  };
 
-  if (mgos_sys_config_get_shelly_mode() == 4) {
-    rgbw_light.reset(new hap::LightBulb(1, FindInput(1), FindOutput(1),
-                                        FindOutput(2), FindOutput(3),
-                                        FindOutput(4), lb_cfg));
-  } else {
-    rgbw_light.reset(new hap::LightBulb(1, FindInput(1), FindOutput(1),
-                                        FindOutput(2), FindOutput(3), nullptr,
-                                        lb_cfg));
+  int mode = mgos_sys_config_get_shelly_mode();
 
-    FindOutput(4)->SetStatePWM(0.0f, "cc");
-  }
+  int ndev = 1;
 
-  if (rgbw_light == nullptr || !rgbw_light->Init().ok()) {
-    return;
+  if (mode == (int) Mode::kCCT) {
+    ndev = 2;
+  } else if (mode == (int) Mode::kWhite) {
+    ndev = 4;
+  } else if (mode == (int) Mode::kRGBpW) {
+    ndev = 2;
   }
 
-  rgbw_light->set_primary(true);
-  mgos::hap::Accessory *pri_acc = (*accs)[0].get();
-  pri_acc->SetCategory(kHAPAccessoryCategory_Lighting);
-  pri_acc->AddService(rgbw_light.get());
-  comps->emplace_back(std::move(rgbw_light));
+  int out_pin = 1;
+  bool first_detatched_input = true;
+  bool is_optional = false;
+
+  for (int i = 0; i < ndev; i++) {
+    lb_cfg = lb_cfgs[i];
+
+    if (mode == (int) Mode::kRGB) {
+      lightbulb_controller.reset(new RGBWController(
+          lb_cfg, FindOutput(1), FindOutput(2), FindOutput(3), nullptr));
+      FindOutput(4)->SetStatePWM(0.0f, "cc");
+    } else if (mode == (int) Mode::kRGBW) {
+      lightbulb_controller.reset(new RGBWController(
+          lb_cfg, FindOutput(1), FindOutput(2), FindOutput(3), FindOutput(4)));
+    } else if (mode == (int) Mode::kCCT) {
+      lightbulb_controller.reset(new CCTController(lb_cfg, FindOutput(out_pin),
+                                                   FindOutput(out_pin + 1)));
+      out_pin += 2;
+      is_optional = true;
+    } else if (mode == (int) Mode::kWhite) {
+      lightbulb_controller.reset(
+          new LightController(lb_cfg, FindOutput(out_pin++)));
+      is_optional = (mgos_sys_config_get_shelly_mode() != (int) Mode::kRGBpW);
+    } else {  // Mode::kRGBpW
+      lightbulb_controller.reset(new RGBWController(
+          lb_cfg, FindOutput(1), FindOutput(2), FindOutput(3), nullptr));
+      mode = (int) Mode::kWhite;  // last bulb is White
+      out_pin += 3;
+    }
+
+    Input *in = FindInput(1);
+    if (i != 0) {
+      in = nullptr;  // support input only for first device
+    }
+
+    hap_light.reset(new hap::LightBulb(
+        i + 1, in, std::move(lightbulb_controller), lb_cfg, is_optional));
 
-  if (lb_cfg->in_mode == 3) {
-    hap::CreateHAPInput(1, mgos_sys_config_get_in1(), comps, accs, svr);
+    if (hap_light == nullptr || !hap_light->Init().ok()) {
```

Comment:
> won't hurt to print init error, if there's one.

from @rojer

Reply:
It's just copy&past from for example the GDO code from the 1 or 2.5.

Reply:
you're right, i shouldn't be holding you to a higher standard than myself :)
so [here goes](https://github.com/mongoose-os-apps/shelly-homekit/commit/77b157af07c6d2d74aab24026cdcd4abc2d58ea4).